crates/accelerate/src/quantum_circuit/intern_context.rs

```diff
@@ -0,0 +1,71 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::exceptions::PyRuntimeError;
+use pyo3::PyResult;
+use std::sync::Arc;
+
+pub type IndexType = u32;
+pub type BitType = u32;
+
+/// A Rust-only data structure (not a pyclass!) for interning
+/// `Vec<BitType>`.
+///
+/// Takes ownership of vectors given to [InternContext.intern]
+/// and returns an [IndexType] index that can be used to look up
+/// an _equivalent_ sequence by reference via [InternContext.lookup].
+#[derive(Clone, Debug)]
+pub struct InternContext {
+    slots: Vec<Arc<Vec<BitType>>>,
+    slot_lookup: HashMap<Arc<Vec<BitType>>, IndexType>,
```

Comment:
These are `Arc<Vec<_>>` just because they're the same `Vec`, right?  Given that they should then be private within this context object, can we remove the sync overhead and make them `Rc<Vec<_>>` instead?  Although I suppose we'd have to revert to `Arc` as soon as this gets exposed to Python space anyway, since Rust wouldn't be happy otherwise.

Reply:
That's correct. And also, I realized earlier today that using `Arc` for these actually gives us decent `clone` semantics for `InternContext`, since `Arc` clones by reference, and thus `InternContext::clone` only ends up cloning the tracking info, while sharing the bit lists. A decent mix of lock-free sharing with scaling that isn't tied to size of the bit lists.

The reason this can't be `Rc` is exactly what you mention. The `Rc`s would be stored in `InternContext`, which is stored in `CircuitData`, which is a a `pyclass`, and thus must be `Send`.